Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/elm/execution/FunctionRefEvaluator.java

```diff
@@ -5,11 +5,16 @@
 
 import org.cqframework.cql.elm.execution.Expression;
 import org.cqframework.cql.elm.execution.FunctionDef;
+import org.cqframework.cql.elm.execution.FunctionRef;
 import org.opencds.cqf.cql.engine.execution.Context;
 import org.opencds.cqf.cql.engine.execution.Variable;
 
 public class FunctionRefEvaluator extends org.cqframework.cql.elm.execution.FunctionRef {
 
+    // For a given instance of the FunctionDefEvaluator, we'll only resolve to one FunctionDef
+    // Even if it's evaluated multiple times in a given ELM graph.
+    private FunctionDef functionDef;
```

Comment:
I'm concerned that this might not always be correct, because the resolution is based on the runtime argument types. Consider:

```cql
define function TestAny(a Integer):
  a

define function TestAny(b String):
  b

define TestAnyFunctionWithInteger:
  TestAny(1)

define TestAnyFunctionWithString:
  TestAny('joe')
```

Currently, this works as expected because of the runtime resolution:

```cql
TestAnyFunctionWithInteger=1
TestAnyFunctionWithString=joe
```
The caching here would have to be expanded to include the resolved signature.

Reply:
Agreed that we need to make some adjustments to workaround some issues but _conceptually_ we _should_ be able to do this, right? The fact that we need to rely on runtime type resolution is actually an issue with the ELM generation. The 2 `FunctionRef` elements that are generated for the 2 separate calls to 2 separate overloads `TestAny` are indistinguishable and we can end up with ambiguous function resolutions.

The resolved signature is included in the `FunctionDesc` class that I added a couple PRs back. We could potentially use those but that doesn't actually save us from the most expensive part of the function evaluation, which is determining the types and compatibility of the function arguments via reflection. So, I think we come via a solution to address that.